### ISSUE TITLE
Fix zero-extension vasms on PPC64

### ISSUE THREAD

**Issue by gut** [PPC64/hhvm#155]
Currently the vasms that should zero-extend small registers to bigger registers are sign-extending (although it works): movzbw, movzbl, movzbq, movzlq.

The following diff should fix this issue:

``` diff
diff --git i/hphp/runtime/vm/jit/vasm-ppc64.cpp w/hphp/runtime/vm/jit/vasm-ppc64.cpp
index a872785..5ce96af 100644
--- i/hphp/runtime/vm/jit/vasm-ppc64.cpp
+++ w/hphp/runtime/vm/jit/vasm-ppc64.cpp
@@ -1313,9 +1313,9 @@ void lowerForPPC64(Vout& v, movtql& inst) { v << copy{inst.s, inst.d}; }
 void lowerForPPC64(Vout& v, movtdb& inst) { v << copy{inst.s, inst.d}; }
 void lowerForPPC64(Vout& v, movtdq& inst) { v << copy{inst.s, inst.d}; }

-// Lower all movzb* to extrb as ppc64 always sign extend the unused bits of reg.
-void lowerForPPC64(Vout& v, movzbl& i)    { v << extrb{i.s, Reg8(i.d)}; }
-void lowerForPPC64(Vout& v, movzbq& i)    { v << extrb{i.s, Reg8(i.d)}; }
+// Lower all movzb* to simple mov* to extract useful part and zero upper bits.
+void lowerForPPC64(Vout& v, movzbl& i)    { v << movb{i.s, Reg8(i.d)}; }
+void lowerForPPC64(Vout& v, movzbq& i)    { v << movb{i.s, Reg8(i.d)}; }
 void lowerForPPC64(Vout& v, movzlq& i)    { v << movl{i.s, Reg32(i.d)}; }

 void lowerForPPC64(Vout& v, srem& i) {
```

However on 3.15.1 I tested and the following 3 tests failed on release mode:
hphp/test/quick/builtin_extension_SimpleXMLIterator.php
hphp/test/quick/fb_dbt.php
hphp/test/quick/recursive_iterator_iterator_empty.php

Max Wang suggested that this failure indicates a bug somewhere else. The data should indeed be zero-extended.

**Comment by racardoso** [comment]
Those tests fails on Debug mode not on Release mode. On Release mode all tests runs fine. Also we got this problem on our next branch too:

Failing tests HHVM-3.15.1:

```
quick/builtin_extension_SimpleXMLIterator.php
quick/fb_dbt.php
quick/json_last_error.php
quick/recursive_iterator_iterator_empty.php
```

Failing tests next:

```
quick/builtin_extension_SimpleXMLIterator.php
quick/debugger/flow_iterator_continue.php
quick/fb_dbt.php
quick/json_last_error.php
quick/recursive_iterator_iterator_empty.php
```

**Comment by racardoso** [comment]
The problem was the instruction `movb` was being removed as a `dead code` since it was considered a trivial nop (if movb r(dst),r(src)) https://github.com/facebook/hhvm/blob/10b4a1aeb38979d8632ab7496ddab390777f6b2d/hphp/runtime/vm/jit/vasm-util.cpp#L32

Later removed by https://github.com/facebook/hhvm/blob/cf16ad5393559aedd14cd61cc5e0f8c94ec21b11/hphp/runtime/vm/jit/vasm-dead.cpp#L383 

On PPC64 the `movb` vasm instruction is mapped to a `rlwinm` instruction that also zero upper bits of register and is used in `movzb` type instructions unlike x86_64. Hence we cannot consider `movb` as a trivial nop on PPC64 even if r(src) = r(dst) since we may alter the register content .

Closed via 4046deb
